### PR TITLE
Changed setting hparams dict to use update()

### LINKED ISSUE #9083: Instead of overwriting `hparams` when resuming from a checkpoint, merge them
## 🚀 Feature
Instead of doing:
```
# overwrite hparams by the given file
checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY] = hparams
```
[Source](https://github.com/PyTorchLightning/pytorch-lightning/blob/e63968ab8867a5aae8ab9d883f186ce379199cff/pytorch_lightning/core/saving.py#L144)

It could be:
```
checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY].update(hparams)
```

### Motivation
I had saved some hyperparameters for logging to tensorboard hparams. However, not all of the variables were saved (since I only wanted a few to show up in the hparams section). When I did:
```
MyLightningModule.load_from_checkpoint(checkpoint_path=pth, hparams_file=hp_file)
```
I then got an error because certain variables that aren't present in the hparams file I specified.

However, if I do:
```
MyLightningModule.load_from_checkpoint(checkpoint_path=pth)
```
then all the variables are loaded from the checkpoint and everything works as expected.

I can also do:
```
MyLightningModule.load_from_checkpoint(
    checkpoint_path=pth, 
    hparams_file=hp_file,
    # extra variables missing from hparams file here
)
```
but the cleanest approach would be to just use the hparams file to update the existing variables in the checkpoint (rather than overwriting everything from the checkpoint).

### Pitch
If you have some variables you want to override, but still want the majority of the variables saved in the checkpoint, it would be desirable to have the hyperparameters merged rather than only use the hyperparameters from the saved file.

### Alternatives
```
checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY].update(hparams)
```

### Additional context
N/A

### PR DESCRIPTION
**Pull request by EricWiener**
## What does this PR do?
When loading a checkpoint and a hparams file, instead of only loading state from the hparams file, will now load from the checkpoint first and then update with values from the hparams file.

Fixes #9083 (see motivation here)

### Does your PR introduce any breaking changes? If yes, please list them.
No.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Always

### CODE REVIEW

File: tests/checkpointing/test_model_checkpoint.py [hunk 1]
```diff
@@ -1248,3 +1248,42 @@ def test_trainer_checkpoint_callback_bool(tmpdir):
...
+    # Make sure that even though y isn't saved in hparams_file,
+    # the model can still be loaded using the y value saved
+    # in the checkpoint.
+    model = TestModel.load_from_checkpoint(
+        checkpoint_path=ckpt_path,
+        hparams_file=hparams_path,
+    )
+
+    assert model.x == 5
+    assert model.y == 6
```
Comment (rohitgr7) [inline tests/checkpointing/test_model_checkpoint.py:1289]:
```suggestion
def test_load_checkpoint_update_hparams(tmpdir):
    """
    Check that old hparams are updated when new hparams_file is provided
    """
    class TestModel(BoringModel):
        def __init__(self, x=None, y=None):
            super().__init__()
            self.save_hyperparameters()

    model_checkpoint = ModelCheckpoint(dirpath=tmpdir, filename="{epoch:02d}")
    trainer = Trainer(
        max_epochs=1,
        default_root_dir=tmpdir,
        limit_train_batches=1,
        limit_val_batches=1,
        callbacks=[model_checkpoint],
        logger=TensorBoardLogger(save_dir=tmpdir),
    )
    model = TestModel(5)
    trainer.fit(model)

    ckpt_path = os.path.join(tmpdir, "epoch=00.ckpt")
    hparams_path = os.path.join(tmpdir, "default", "version_0", "hparams.yaml")
    assert os.path.exists(hparams_path)
    assert model.hparams.x == 5
    assert model.hparams.y is None

    with open(hparams_path, 'w') as f:
        yaml.dump({'y': 10}, f)

    # Make sure that even though x isn't saved in new hparams_file,
    # the model can still be loaded using the x value saved
    # in the checkpoint.
    model = TestModel.load_from_checkpoint(
        checkpoint_path=ckpt_path,
        hparams_file=hparams_path,
    )

    assert model.hparams.x == 5
    assert model.hparams.y == 10
```
the test passes on master since save_hyperparameters saves x and y both in the checkpoint hparams. For this one, I think providing a new hparams file which only contains some missing hparams would be a better way to test this. You might need to import yaml up there. Also, I think this test is better suited for `tests/models/test_hparams`.

File: pytorch_lightning/core/saving.py [hunk 2]
```diff
@@ -141,8 +141,11 @@ def load_from_checkpoint(
 
             hparams["on_gpu"] = False
 
-            # overwrite hparams by the given file
-            checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY] = hparams
+            # overwrite hparams by the given file (if they exist in the checkpoint)
+            if cls.CHECKPOINT_HYPER_PARAMS_KEY in checkpoint:
+                checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY].update(hparams)
+            else:
+                checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY] = hparams
```
Comment (rohitgr7) [inline pytorch_lightning/core/saving.py:148]: just wondering in what case do we not have `cls.CHECKPOINT_HYPER_PARAMS_KEY` already present in the checkpoint while loading. any idea??

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (EricWiener) [reply tests/checkpointing/test_model_checkpoint.py:1289]: Thanks for the feedback! However, I'm pretty sure `save_hyperparameters` would only save properties defined before it is called. I had verified locally (on PL `1.4.4`) and doing:

```
self.x = x
self.save_hyperparameters()
self.y = y
```

results in a `hparams.yaml` that only contains:

```
x: 5
```

Is this behavior changed in 1.5.0?

Regardless, I think your solution is better, just wanted to make sure my assumption about `save_hyperparameters` is correct?

File: pytorch_lightning/core/saving.py (hunk 2 above)
Comment (EricWiener) [reply pytorch_lightning/core/saving.py:148]: One of the CI tests failed earlier because it wasn't present, but I forgot which one (and the CI doesn't show for my initial commit anymore on this branch)

Comment (EricWiener) [reply pytorch_lightning/core/saving.py:148]: Checking CI again with: https://github.com/PyTorchLightning/pytorch-lightning/pull/9180

Comment (EricWiener) [reply pytorch_lightning/core/saving.py:148]: Here's failed test: https://github.com/PyTorchLightning/pytorch-lightning/pull/9180/checks?check_run_id=3452345635#step:6:2405

File: tests/checkpointing/test_model_checkpoint.py (hunk 1 above)
Comment (rohitgr7) [reply tests/checkpointing/test_model_checkpoint.py]: AFAIK `save_hyperparameters()` doesn't care what you save as a state variable and where you save it. It just captures all the args and kwargs and saves them. You can try reverting your changes in the codebase and run your current test, it will pass. I tried it on master and it's saving both x and y even when I don't attach them to `self`. You can check a few examples here: https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#save-hyperparameters. none of them assign it to the state but save_hyperparameters still saves them.

File: pytorch_lightning/core/saving.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/core/saving.py:148]: hmm.. looks like another bug to me.. need to explore more. but yeah changes are fine since user can remove the key by themselves this takes care of that situation as well.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by ethanwharris — APPROVED** [review]

File: tests/models/test_hparams.py [hunk 3]
```diff
@@ -778,3 +780,42 @@ def test_colliding_hparams(tmpdir):
...
+    )
+    model = TestModel(5)
+    trainer.fit(model)
+    ckpt_path = os.path.join(tmpdir, "epoch=00.ckpt")
+    hparams_path = os.path.join(tmpdir, "default", "version_0", "hparams.yaml")
+    assert os.path.exists(hparams_path)
+    assert model.hparams.x == 5
+    assert model.hparams.y is None
+    with open(hparams_path, "w") as f:
+        yaml.dump({"y": 10}, f)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/models/test_hparams.py:812]: > Depending on the use case, requiring this might suggest that x=5 should be the default value for the model.

Very good point and exactly my concern I tried to express in the linked issue :) 
Printing a warning could mitigate this, however the danger with warnings over errors is that users miss them or outright ignore :)

File: pytorch_lightning/core/saving.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/core/saving.py:148]:
```suggestion
            if cls.CHECKPOINT_HYPER_PARAMS_KEY not in checkpoint:
                checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY] = {}
            checkpoint[cls.CHECKPOINT_HYPER_PARAMS_KEY].update(hparams)    
```

**Review by Borda — APPROVED** [review]

File: tests/models/test_hparams.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EricWiener) [reply tests/models/test_hparams.py:812]: Thanks for the feedback. I'm gonna go ahead and close the issue + PR
